Dir.home not a reliable way to obtain user's home directory

# Summary

When the relevant "NETRC" and/or "HOME" environment variables are not present, the `Netrc` module attempts to fallback on using the user's home directory as returned by Ruby's `Dir.home` method. That *ought* to be *the* reliable way to do it, but it was recently found that it does not work for processes that are not the (grand)child of a login process. Details on why this is are below.

On Unix-like systems, when `Dir.home` chokes, `Netrc` should try to obtain the user's home directory from the password database using the `Ext.getpwuid(...)` method. This can succeed because it can be invoked with the value of `Process.uid` provided as an argument. Those calls, of course, are just relatively thin Ruby wrappers around `getuid(2)` (which should never fail) and `getpwuid(3)` (or `getpwuid_r(3)`); neither is dependent upon whether the process is a login descendant.

Even if most uses of the `Netrc` module would be in processes that *are* (grand)children of login processes, the unit tests exercise the module's fallback code that attempts to deal with scenarios where the more common "NETRC" and/or "HOME" environment variables are not present. That unit test code fails when run in a non-login-descendant process (as is the case on with GitHub Actions integration running on an 'ubuntu-latest' host).

Perhaps more importantly, though, the actual runtime behavior of the `Netrc` module is different when run as a non-login-descended process. As it regards this issue, it ought to run the same either way.

The best long-term fix would be for `Dir.home` to be enhanced in the upstream Ruby code base to remove the discrepancy, but even if that can be brought about it would still take a long time for updated versions of Ruby to make their way out into the world. For that reason, I think it would be nice to have a workaround for the `Dir.home` issue in `Netrc`.


# Background

When @geemus configured the project with continuous integration on GitHub Actions [the other day](https://github.com/heroku/netrc/pull/47#issuecomment-611741058), the unit tests -- which worked locally on my machine -- failed thus:

```
2020-04-09T19:30:32.5010432Z .........E......................................
2020-04-09T19:30:32.5016905Z 
2020-04-09T19:30:32.5021772Z Finished in 0.106068s, 452.5405 runs/s, 697.6666 assertions/s.
2020-04-09T19:30:32.5027512Z 
2020-04-09T19:30:32.5035730Z   1) Error:
2020-04-09T19:30:32.5036590Z TestNetrc#test_missing_environment:
2020-04-09T19:30:32.5039408Z ArgumentError: couldn't find login name -- expanding `~'
2020-04-09T19:30:32.5041265Z     /home/runner/work/netrc/netrc/test/test_netrc.rb:256:in `home'
2020-04-09T19:30:32.5042455Z     /home/runner/work/netrc/netrc/test/test_netrc.rb:256:in `test_missing_environment'
2020-04-09T19:30:32.5043093Z 
2020-04-09T19:30:32.5043510Z 48 runs, 74 assertions, 0 failures, 1 errors, 0 skips
2020-04-09T19:30:32.5050891Z ##[error]Process completed with exit code 1.
```

The call path there bottoms out as Ruby's `Dir.home` tries to obtain the user's home directory from the `File` module (this is from `file.c` in the ruby source tree):

```C
    #ifndef _WIN32
    VALUE
    rb_default_home_dir(VALUE result)
    {
        const char *dir = getenv("HOME");

    #if defined HAVE_PWD_H
        if (!dir) {
        const char *login = getlogin();
        if (login) {
            struct passwd *pw = getpwnam(login);
            if (pw) {
            copy_home_path(result, pw->pw_dir);
            endpwent();
            return result;
            }
            endpwent();
            rb_raise(rb_eArgError, "couldn't find HOME for login `%s' -- expanding `~'",
                 login);
        }
        else {
            rb_raise(rb_eArgError, "couldn't find login name -- expanding `~'");
        }
        }
    #endif
        if (!dir) {
        rb_raise(rb_eArgError, "couldn't find HOME environment -- expanding `~'");
        }
        return copy_home_path(result, dir);
    }
```

The `getlogin()` function is arguably not the right way for Ruby to go about obtaining that information in that context, as it will return `NULL` for a process that is not a descendant of login. An approach that would work for broader set of scenarios would be to call `getuid()`, and then `getpwuid_r(...)`, as outlined above.

# Debug cookbook

For others who might encounter similar issues, I'll note that the fingerprint for this on GNU/Linux hosts is a discrepancy between the values reported by `id(1)` and the `loginuid` attribute:

```bash
    $ /usr/bin/id
    uid=1001(runner) gid=115(docker) groups=115(docker)

    $ /usr/bin/id --user
    1001

    /usr/bin/getent passwd 1001
    runner:x:1001:115:,,,:/home/runner:/bin/bash

    $ cat /proc/self/loginuid
    4294967295
```

In the particular case of GitHub Actions, you can arrange for an "action" to run a sequence of shell commands (to probe the environment) and then view their output via the GitHub Actions "View raw logs" option.

```yaml
name: Whatever

on:
    push:
        branches: [ ads/filehome-vs-login-muckage ]

    pull_request:
        branches: [ ads/filehome-vs-login-muckage ]

jobs:
    mybuild:

        runs-on: ubuntu-latest

        steps:
            - name: Run my debugging commands
              run: |
                  echo NOOP set -e
                  printf '%s\n' 'ALJUNK: go' 1>&2
                  set -x
                  :
                  : Note that basic environment looks sane
                  :
                  /usr/bin/whoami             1>&2 || true
                  /usr/bin/id                 1>&2 || true
                  /usr/bin/id --user          1>&2 || true
                  /usr/bin/getent passwd $(/usr/bin/id --user) 1>&2 || true
                  :
                  : This will be 4294967295  a.k.a. -1  for a non-login-descendant process
                  cat /proc/self/loginuid     1>&2 || true
                  :
                  :
                  /usr/bin/ruby2.5 -e 'print "process uid: #{Process.uid}\n";' 1>&2 || true
                  /usr/bin/ruby2.5 -r etc -e 'print "can we load the etc module? Record: #{Etc.getpwuid( Process.uid )}\n";' 1>&2 || true
                  :
                  : When HOME is set, as it is by default, Dir.home works without a problem
                  echo "$HOME"                     1>&2 || true
                  /usr/bin/ruby2.5 -e 'Dir.home;'  1>&2 || true
                  :
                  : When HOME is not set....kaboom!
                  ( set -x; unset HOME; /usr/bin/ruby2.5 -e 'Dir.home;'; echo $? ) 1>&2 || true
```

Here's the output of that "action" from GitHub Actions's raw logs output. Notice that `$HOME` is set by default, and the value of the `loginuid` attribute is `-1` (`4294967295`). With that arrangement you can see that Ruby's `Dir.home` works fine. But when we then unset the `$HOME` environment variable in the subshell, it breaks. That's the same scenario the `Netrc` unit tests are creating, and hence the reason for their failure currently when running on GitHub Actions.

```
...
2020-04-13T04:13:12.9458811Z NOOP set -e
2020-04-13T04:13:12.9541831Z ALJUNK: go
2020-04-13T04:13:12.9542252Z + :
2020-04-13T04:13:12.9542463Z + : Note that basic environment looks sane
2020-04-13T04:13:12.9542662Z + :
2020-04-13T04:13:12.9542803Z + /usr/bin/whoami
2020-04-13T04:13:12.9542989Z runner
2020-04-13T04:13:12.9543168Z + /usr/bin/id
2020-04-13T04:13:12.9543323Z uid=1001(runner) gid=115(docker) groups=115(docker)
2020-04-13T04:13:12.9545310Z + /usr/bin/id --user
2020-04-13T04:13:12.9545567Z 1001
2020-04-13T04:13:12.9545947Z ++ /usr/bin/id --user
2020-04-13T04:13:12.9549178Z + /usr/bin/getent passwd 1001
2020-04-13T04:13:12.9556052Z runner:x:1001:115:,,,:/home/runner:/bin/bash
2020-04-13T04:13:12.9559902Z + :
2020-04-13T04:13:12.9560979Z + : This will be 4294967295 a.k.a. -1 for a non-login-descendant process
2020-04-13T04:13:12.9561394Z + cat /proc/self/loginuid
2020-04-13T04:13:12.9561543Z 4294967295+ :
2020-04-13T04:13:12.9562307Z + :
2020-04-13T04:13:12.9562845Z + /usr/bin/ruby2.5 -e 'print "process uid: #{Process.uid}\n";'
2020-04-13T04:13:12.9564509Z /usr/lib/ruby/vendor_ruby/rubygems/defaults/operating_system.rb:10: warning: constant Gem::ConfigMap is deprecated
2020-04-13T04:13:12.9565605Z /usr/lib/ruby/vendor_ruby/rubygems/defaults/operating_system.rb:10: warning: constant Gem::ConfigMap is deprecated
2020-04-13T04:13:12.9566434Z /usr/lib/ruby/vendor_ruby/rubygems/defaults/operating_system.rb:29: warning: constant Gem::ConfigMap is deprecated
2020-04-13T04:13:12.9567489Z /usr/lib/ruby/vendor_ruby/rubygems/defaults/operating_system.rb:30: warning: constant Gem::ConfigMap is deprecated
2020-04-13T04:13:12.9567880Z process uid: 1001
2020-04-13T04:13:12.9568613Z + /usr/bin/ruby2.5 -r etc -e 'print "can we load the etc module? Record: #{Etc.getpwuid( Process.uid )}\n";'
2020-04-13T04:13:12.9568905Z /usr/lib/ruby/vendor_ruby/rubygems/defaults/operating_system.rb:10: warning: constant Gem::ConfigMap is deprecated
2020-04-13T04:13:12.9569657Z /usr/lib/ruby/vendor_ruby/rubygems/defaults/operating_system.rb:10: warning: constant Gem::ConfigMap is deprecated
2020-04-13T04:13:12.9570179Z /usr/lib/ruby/vendor_ruby/rubygems/defaults/operating_system.rb:29: warning: constant Gem::ConfigMap is deprecated
2020-04-13T04:13:12.9570619Z /usr/lib/ruby/vendor_ruby/rubygems/defaults/operating_system.rb:30: warning: constant Gem::ConfigMap is deprecated
2020-04-13T04:13:12.9570948Z can we load the etc module? Record: #<struct Etc::Passwd name="runner", passwd="x", uid=1001, gid=115, gecos=",,,", dir="/home/runner", shell="/bin/bash">
2020-04-13T04:13:12.9571393Z + :
2020-04-13T04:13:12.9571652Z + : When HOME is set, as it is by default, Dir.home works without a problem
2020-04-13T04:13:12.9571861Z + echo /home/runner
2020-04-13T04:13:12.9572055Z /home/runner
2020-04-13T04:13:12.9573063Z + /usr/bin/ruby2.5 -e 'Dir.home;'
2020-04-13T04:13:12.9573432Z /usr/lib/ruby/vendor_ruby/rubygems/defaults/operating_system.rb:10: warning: constant Gem::ConfigMap is deprecated
2020-04-13T04:13:12.9574008Z /usr/lib/ruby/vendor_ruby/rubygems/defaults/operating_system.rb:10: warning: constant Gem::ConfigMap is deprecated
2020-04-13T04:13:12.9574527Z /usr/lib/ruby/vendor_ruby/rubygems/defaults/operating_system.rb:29: warning: constant Gem::ConfigMap is deprecated
2020-04-13T04:13:12.9575186Z /usr/lib/ruby/vendor_ruby/rubygems/defaults/operating_system.rb:30: warning: constant Gem::ConfigMap is deprecated
2020-04-13T04:13:12.9575411Z + :
2020-04-13T04:13:12.9575963Z + : When HOME is not 'set....kaboom!'
2020-04-13T04:13:12.9576633Z + set -x
2020-04-13T04:13:12.9577073Z + unset HOME
2020-04-13T04:13:12.9577799Z + /usr/bin/ruby2.5 -e 'Dir.home;'
2020-04-13T04:13:12.9578212Z /usr/lib/ruby/vendor_ruby/rubygems/defaults/operating_system.rb:10: warning: constant Gem::ConfigMap is deprecated
2020-04-13T04:13:12.9578592Z /usr/lib/ruby/vendor_ruby/rubygems/defaults/operating_system.rb:10: warning: constant Gem::ConfigMap is deprecated
2020-04-13T04:13:12.9578968Z /usr/lib/ruby/vendor_ruby/rubygems/defaults/operating_system.rb:29: warning: constant Gem::ConfigMap is deprecated
2020-04-13T04:13:12.9579301Z /usr/lib/ruby/vendor_ruby/rubygems/defaults/operating_system.rb:30: warning: constant Gem::ConfigMap is deprecated
2020-04-13T04:13:12.9581745Z -e:1:in `home': couldn't find login name -- expanding `~' (ArgumentError)
2020-04-13T04:13:12.9582151Z 	from -e:1:in `<main>'
2020-04-13T04:13:12.9582575Z + echo 1
2020-04-13T04:13:12.9582956Z 1
...
```

# See also
For the original context that lead to this issue, please see the discussion for PR #47 (related to issue #46).

--- a/lib/netrc.rb
+++ b/lib/netrc.rb
@@ -24,6 +24,32 @@ class Netrc
 
     (home && File.readable?(home)) ? home : Dir.pwd
   rescue ArgumentError
+
+    unless WINDOWS
+      # Ruby 2.5.x through 2.7.1 (and probably other versions) raises an
+      # ArgumentError when the 'HOME' env var is not set and the process is
+      # not a subprocess of a login session. That is, it tries to navigate
+      # into the password database using the username obtained from
+      # getlogin(), whichs will return NULL in such a circumstance. It would
+      # be better if it instead tried to do it via the UID obtained from
+      # getuid(), which always succeeds. Nevertheless, we'll work around the
+      # problem by doing that ourselves here.
+      begin
+        require 'etc'
+      rescue LoadError
+        warn "HOME is not set, and the 'etc' module not available; using pwd as home\n"
+        return Dir.pwd
+      end
+
+      passwd_record = Etc.getpwuid(Process.uid);
+      unless passwd_record
+        warn "Record for uid #{Process.uid} not found in the password database; using pwd as home\n"
+        return Dir.pwd
+      end
+
+      return passwd_record.dir
+    end
+
     return Dir.pwd
   end
 

--- a/test/test_netrc.rb
+++ b/test/test_netrc.rb
@@ -20,6 +20,38 @@ class TestNetrc < Minitest::Test
     Dir.glob('data/*.netrc').each{|f| File.chmod(0644, f)}
   end
 
+  # Helper for obtaining a value for the user's home directory from the
+  # password database (querying it by uid), else falling back on using the
+  # pwd. This is analogous to what the Netrc.home_path method does on
+  # Unix-like systems when neither 'NETRC' nor 'HOME' are defined, if the
+  # process is not a child (or grandchild) of a login session, or if the
+  # user's actual home directory is not readable for some reason.
+  #
+  def fallback_homedir_else_pwd
+    begin
+      require 'etc'
+    rescue LoadError
+      # Without the 'Etc' module we are unable to query the password database
+      return Dir.pwd
+    end
+
+    # Note that Process.uid returns the value of getuid() which (on Linux) may
+    # be different from the value of /proc/self/loginuid (e.g., for a process
+    # that is not a (grand)child of a login process). That is important here
+    # because we should be able to obtain the record from the password
+    # database, even if Ruby's built-in 'Dir.home' cannot (because it uses
+    # getlogin() and fails in that scenario, at least through Ruby 2.7.1).
+    #
+    passwd_record = Etc.getpwuid(Process.uid)
+    unless passwd_record
+      return Dir.pwd  # Record for uid not available for some reason
+    end
+
+    return passwd_record.dir if File.readable?(passwd_record.dir)
+
+    return Dir.pwd
+  end
+
   def test_parse_empty
     pre, items = Netrc.parse(Netrc.lex([]))
     assert_equal("", pre)
@@ -245,20 +277,40 @@ class TestNetrc < Minitest::Test
     end
   end
 
+  # The precedence order for finding the user's netrc file is the first
+  # /readable/ name from the following list:
+  #
+  #     1. Use NETRC, if set
+  #     2. Use "$HOME/.netrc", if HOME is set
+  #     3. Use Dir.home, if possible (only works where getlogin() is non-NULL)
+  #     4. (MS Windows only) Combine HOMEDRIVE and HOMEPATH, if both are set
+  #     5. (MS Windows only) Use USERPROFILE, if set
+  #     6. (Non-MS Windows only) Use pw_dir field from password database, if available
+  #     7. Dir.pwd  (is not checked for readability)
+  #
+  # This test exercises the behavior when none of the above listed environment
+  # variables are present and/or set.
+  #
   def test_missing_environment
-    nil_home = nil
-    ENV["HOME"], nil_home = nil_home, ENV["HOME"]
 
-    # If user's home directory can be obtained via getpwnam(3) AND is
-    # readable, then the $HOME environment variable is not considered for
-    # building the Netrc default_path, and no fallback directory is
-    # referenced.
-    dflt_dir = Dir.respond_to?(:home) && File.readable?(Dir.home) \
-             ? Dir.home : Dir.pwd
+    envkeys = %w(NETRC HOME HOMEDRIVE HOMEPATH USERPROFILE)
+    envhold = {}
+
+    envkeys.each do |ekey|
+      if ENV.has_key?(ekey)
+        envhold[ekey] = ENV[ekey]
+        ENV.delete(ekey)
+      end
+    end
+
+    dflt_dir = self.fallback_homedir_else_pwd
 
     assert_equal File.join(dflt_dir, '.netrc'), Netrc.default_path
   ensure
-    ENV["HOME"], nil_home = nil_home, ENV["HOME"]
+    # ENV obj is not really a hash; use poor man's manual merge!(...)
+    envhold.each do |holdkey, holdval|
+      ENV[holdkey] = holdval
+    end
   end
 
   def test_netrc_environment_variable

